### PR TITLE
nixos/rspamd-trainer: init; rspamd-trainer: init at unstable-2023-11-27

### PR DESCRIPTION
**Pull request by onny**
## Description of changes

[Rspamd-trainer](https://gitlab.com/onlime/rspamd-trainer) script and timer which periodically checks a specific inbox for spam/ham mails which it passes to rspamd for training

```
services.rspamd-trainer = {
  enable = true;
  settings = {
    HOST = "localhost";
    USERNAME = "spam@example.com";
    PASSWORD = "test123";
    INBOXPREFIX = "INBOX/";
  };
  secrets = [
    # Do not use this in production. This will make passwords
    # world-readable in the Nix store
    "${pkgs.writeText "secrets" ''
      PASSWORD = test123
    ''}"
  ];
};
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/mail/rspamd-trainer.nix [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+  config = mkIf cfg.enable {
+
+    systemd = {
+      services.rspamd-trainer = {
+        description = "Spam/ham trainer for rspamd";
+        serviceConfig = {
+          ExecStart = "${pkgs.rspamd-trainer}/bin/rspamd-trainer";
+          WorkingDirectory = "/var/lib/rspamd-trainer";
+          StateDirectoryMode = "750";
```
Comment (happysalada) [inline nixos/modules/services/mail/rspamd-trainer.nix:53]: why would you need this ?

File: nixos/modules/services/mail/rspamd-trainer.nix [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+          StateDirectory = [
```
Comment (happysalada) [inline nixos/modules/services/mail/rspamd-trainer.nix:54]: why do you need 2 different ones ? Wouldn't the other be the LogDirectory ?

**Comment by happysalada** [comment]
I have 2 questions but otherwise this looks good

**Comment by onny** [comment]
Thank you @happysalada for reviewing. Fixed the issues and tests also seems to work now :)

**Comment by happysalada** [comment]
fyi, having statedirectory including /log is a little weird.
it would seem that the ideal solution would be somewhat to use statedirectory = "rspamd-trainer" and a separate LogDirectory for the logs.

I understand that working software is better than perfect software, so I'll just leave it here in case you have time for future fixes.

**Pull request merged**